share-coordinator/src/test/java/org/apache/kafka/coordinator/share/ShareCoordinatorRecordSerdeTest.java

```diff
@@ -0,0 +1,249 @@
+    }
+
+    @Test
+    public void testDeserializeWithKeyEmptyBuffer() {
+        ByteBuffer keyBuffer = ByteBuffer.allocate(0);
+        ByteBuffer valueBuffer = ByteBuffer.allocate(64);
+
+        RuntimeException ex =
+                assertThrows(RuntimeException.class,
+                        () -> serde.deserialize(keyBuffer, valueBuffer));
+        assertEquals("Could not read version from key's buffer.", ex.getMessage());
+    }
+
+    @Test
+    public void testDeserializeWithValueEmptyBuffer() {
+        ApiMessageAndVersion key = new ApiMessageAndVersion(
+                new ShareSnapshotKey()
+                        .setGroupId("foo")
+                        .setTopicId(Uuid.randomUuid())
+                        .setPartition(1),
+                ShareCoordinator.SHARE_SNAPSHOT_RECORD_KEY_VERSION
+        );
+        ByteBuffer keyBuffer = MessageUtil.toVersionPrefixedByteBuffer(key.version(), key.message());
+
+        ByteBuffer valueBuffer = ByteBuffer.allocate(0);
+
+        RuntimeException ex =
+                assertThrows(RuntimeException.class,
+                        () -> serde.deserialize(keyBuffer, valueBuffer));
+        assertEquals("Could not read version from value's buffer.", ex.getMessage());
+    }
+
+    @Test
+    public void testDeserializeWithInvalidKeyBytes() {
+        ByteBuffer keyBuffer = ByteBuffer.allocate(2);
+        keyBuffer.putShort((short) 0);
+        keyBuffer.rewind();
+
+        ByteBuffer valueBuffer = ByteBuffer.allocate(2);
+        valueBuffer.putShort((short) 0);
+        valueBuffer.rewind();
+
+        RuntimeException ex =
+                assertThrows(RuntimeException.class,
+                        () -> serde.deserialize(keyBuffer, valueBuffer));
+        assertTrue(ex.getMessage().startsWith("Could not read record with version 0 from key's buffer due to"),
+                ex.getMessage());
+    }
+
+    @Test
+    public void testDeserializeWithInvalidValueBytes() {
+        ApiMessageAndVersion key = new ApiMessageAndVersion(
+                new ShareSnapshotKey()
+                        .setGroupId("foo")
+                        .setTopicId(Uuid.randomUuid())
+                        .setPartition(1),
+                ShareCoordinator.SHARE_SNAPSHOT_RECORD_KEY_VERSION
+        );
+        ByteBuffer keyBuffer = MessageUtil.toVersionPrefixedByteBuffer(key.version(), key.message());
+
+        ByteBuffer valueBuffer = ByteBuffer.allocate(2);
+        valueBuffer.putShort((short) 0);
+        valueBuffer.rewind();
+
+        RuntimeException ex =
+                assertThrows(RuntimeException.class,
+                        () -> serde.deserialize(keyBuffer, valueBuffer));
+        assertTrue(ex.getMessage().startsWith("Could not read record with version 0 from value's buffer due to"),
+                ex.getMessage());
+    }
+
+    @Test
+    public void testDeserializeAllRecordTypes() {
+        roundTrip((short) 0, new ShareSnapshotKey(), new ShareSnapshotValue());
```

Comment:
Does this test ensure that value passed is also verified against the key i.e. if I write mismatched key val `roundTrip((short) 0, new ShareSnapshotKey(), new ShareUpdateValue());` then will it fail?

Reply:
No, it does not. This test is just checking the deserialise method. The message is converted to bytes and then deserialized to check if it is same as original message.
There is no validation on the data or version in this process.

In the actual impl we rely only on the version field of the incoming byte buffer to determine what type of record we want to form.

Reply:
I think this test does check about the keys and passed number identifier, but not the values, correct? I am not sure what we can test with passed value message in this test. You might want to write the comment on top for clarification.

Reply:
No it does not check that - we can pass any version and any key value pair. It will work. Same for group coord counderpart.